### ISSUE TITLE
HTTPS is slow with AWS ELB

### ISSUE THREAD

**Issue by BH4EHN** [square/okhttp#5314]
OkHttp version: 3.12 - 3.14 (only these version tested)
Android version: Android 5 - 8 (only these version tested)

I have a service running on AWS ElasticBeanstalk(EB) and balanced by AWS ElasticLoadBalancer(ELB) handles user uploads. HTTP works totally fine across android/ios/web/electron. Some days ago, I add HTTPS to ELB, each platform works just like before except Android with okhttp have a really low transfer speed both upload and download.
The [SSL Security Policies of ELB](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/create-https-listener.html) is configured use `FS-2018-06` and `2016-08` both have this issue (only test these two).
Currently this part code is replace by HttpsUrlConnection and have a similar speed to other platform.

The upload part code is quite simple
``` Kotlin
val rawFile = File(dataDir, "<filename>.zip")
val httpUrl = HttpUrl.get("https://<domain_to_service>")
    .newBuilder()
    .addQueryParameter("<foo1>", "<bar1>")
    .addQueryParameter("<foo2>", "<bar2>")
    .build()
val okHttpClient = OkHttpClient.Builder()
    .build()
val request = Request.Builder()
    .url(httpUrl)
    .post(RequestBody.create(MediaType.get("application/zip"), rawFile))
    .build()
okHttpClient.newCall(request).enqueue(object: Callback {
    ...notify user transfer over...
})
```

Events, I POST a 128MB file:
```
I/System.out: 0.000 callStart
I/System.out: 0.019 dnsStart
D/libc: [NET] android_getaddrinfofornet+,hn 31(0x75736572646174),sn(),hints(known),family 0,flags 4
    [NET] android_getaddrinfofornet-, err=8
    [NET] android_getaddrinfofornet+,hn 31(0x75736572646174),sn(),hints(known),family 0,flags 1024
    [NET] android_getaddrinfofornet-, pass to proxy
    [NET] android_getaddrinfo_proxy+
D/libc: [NET] android_getaddrinfo_proxy get netid:0
D/libc: [NET] android_getaddrinfo_proxy-, success
I/System.out: 0.297 dnsEnd
I/System.out: 0.302 connectStart
I/System.out: 0.596 secureConnectStart
I/System.out: 1.342 secureConnectEnd
I/System.out: 1.359 connectEnd
    1.359 connectionAcquired
I/System.out: 8.229 requestHeadersStart
I/System.out: 8.241 requestHeadersEnd
    8.242 requestBodyStart
I/System.out: 637.926 requestBodyEnd
    637.926 responseHeadersStart
I/System.out: 646.833 responseHeadersEnd
    646.833 responseBodyStart
    646.839 responseBodyEnd
    646.840 connectionReleased
    646.840 callEnd
```

Network in Profiler, mostly at 0.4MB/s, but on other device same network is at least 3MB/s:
[first 30 seconds](https://imgur.com/hZNj7IH)

**Comment by yschimke** [comment]
Do you have a test URL I could hit?  Happy for you to DM me @yschimke

**Comment by BH4EHN** [comment]
@yschimke I build a test environment with almost the same server configure but no actual server code. I test use the same code above get only 400KB/s upload speed and NodeJS axios I have 3MB/s upload speed.

Test url is here: https://a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/
If the certificate is issued by Amazon, you are hitting the right server.

**Comment by yschimke** [comment]
I haven't dug in much.  FWIW I tested with an OkHttp test client and curl.  with both I was getting around 600KB/s upload from my laptop at home

```
curl -F 'data=@/tmp/file.txt' -vvv https://a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/
```

```
okurl -d @/tmp/file.txt --debug https://a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/
09:09:02.620	url https://a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/
09:09:02.627	Request Request{method=POST, url=https://a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/, tags={class com.baulsupp.okurl.credentials.Token=TokenSet(name=default)}}
09:09:02.685	Dns (a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev): a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/54.172.216.54, a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/52.201.24.243, a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/52.44.133.234
09:09:03.162	>> CONNECTION 505249202a20485454502f322e300d0a0d0a534d0d0a0d0a
09:09:03.164	>> 0x00000000     6 SETTINGS
09:09:03.164	>> 0x00000000     4 WINDOW_UPDATE
09:09:03.166	<< 0x00000000    18 SETTINGS
09:09:03.167	--> POST https://a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/ h2
09:09:03.168	>> 0x00000000     0 SETTINGS      ACK
09:09:03.168	Content-Type: application/x-www-form-urlencoded
09:09:03.168	<< 0x00000000     4 WINDOW_UPDATE
09:09:03.169	Content-Length: 104857600
09:09:03.169	User-Agent: okurl/dev
09:09:03.169	Host: a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev
09:09:03.170	Connection: Keep-Alive
09:09:03.170	Accept-Encoding: gzip
09:09:03.170	Accept-Encoding: br
09:09:03.170	--> END POST
09:09:03.385	Matching interceptor: null
09:09:03.391	>> 0x00000003    91 HEADERS       END_HEADERS
09:09:03.415	<< 0x00000000     0 SETTINGS      ACK
09:09:03.527	>> 0x00000003 65536 DATA
09:09:03.741	<< 0x00000003     4 WINDOW_UPDATE
09:09:03.742	>> 0x00000003  8192 DATA
09:09:03.847	<< 0x00000003     4 WINDOW_UPDATE
09:09:03.848	>> 0x00000003  8192 DATA
09:09:03.849	<< 0x00000003     4 WINDOW_UPDATE
09:09:03.850	>> 0x00000003  8192 DATA
09:09:03.947	<< 0x00000003     4 WINDOW_UPDATE
09:09:03.947	>> 0x00000003  8192 DATA
09:09:03.956	<< 0x00000003     4 WINDOW_UPDATE
09:09:03.957	>> 0x00000003  8192 DATA
09:09:03.957	<< 0x00000003     4 WINDOW_UPDATE
09:09:03.958	>> 0x00000003  8192 DATA
09:09:03.959	<< 0x00000003     4 WINDOW_UPDATE
09:09:03.959	>> 0x00000003  8192 DATA
09:09:03.960	<< 0x00000003     4 WINDOW_UPDATE
```

I am surprised that we end up sending 8KB chunks, this is probably the limiting factor. I'll take a look later today if I have some free time.

**Comment by yschimke** [comment]
Comparison with nghttp2 sending 16KB frames

```
nghttp -nv -d /tmp/file.txt https://a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev/
[  0.136] Connected
The negotiated protocol: h2
[  0.356] recv SETTINGS frame <length=18, flags=0x00, stream_id=0>
          (niv=3)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):128]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65536]
          [SETTINGS_MAX_FRAME_SIZE(0x05):16777215]
[  0.356] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=0>
          (window_size_increment=2147418112)
[  0.356] send SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
[  0.356] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.356] send PRIORITY frame <length=5, flags=0x00, stream_id=3>
          (dep_stream_id=0, weight=201, exclusive=0)
[  0.356] send PRIORITY frame <length=5, flags=0x00, stream_id=5>
          (dep_stream_id=0, weight=101, exclusive=0)
[  0.356] send PRIORITY frame <length=5, flags=0x00, stream_id=7>
          (dep_stream_id=0, weight=1, exclusive=0)
[  0.356] send PRIORITY frame <length=5, flags=0x00, stream_id=9>
          (dep_stream_id=7, weight=1, exclusive=0)
[  0.356] send PRIORITY frame <length=5, flags=0x00, stream_id=11>
          (dep_stream_id=3, weight=1, exclusive=0)
[  0.356] send HEADERS frame <length=71, flags=0x24, stream_id=13>
          ; END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=11, weight=16, exclusive=0)
          ; Open new stream
          :method: POST
          :path: /
          :scheme: https
          :authority: a9828e89-fdee-4924-9d1a-b5adb6424e92.fungo.dev
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.39.1
          content-length: 104857600
[  0.356] send DATA frame <length=16384, flags=0x00, stream_id=13>
[  0.357] send DATA frame <length=16384, flags=0x00, stream_id=13>
[  0.357] send DATA frame <length=16384, flags=0x00, stream_id=13>
[  0.357] send DATA frame <length=16384, flags=0x00, stream_id=13>
[  0.678] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.679] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=13>
          (window_size_increment=16171)
[  0.679] send DATA frame <length=16171, flags=0x00, stream_id=13>
[  0.777] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=13>
          (window_size_increment=16375)
[  0.777] send DATA frame <length=16375, flags=0x00, stream_id=13>
[  0.795] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=13>
          (window_size_increment=16375)
[  0.795] send DATA frame <length=16375, flags=0x00, stream_id=13>
[  0.881] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=13>
          (window_size_increment=16615)
[  0.882] send DATA frame <length=16384, flags=0x00, stream_id=13>
[  0.882] send DATA frame <length=231, flags=0x00, stream_id=13>
[  0.898] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=13>
          (window_size_increment=16171)
[  0.898] send DATA frame <length=16171, flags=0x00, stream_id=13>
[  0.906] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=13>
          (window_size_increment=16375)
[  0.906] send DATA frame <length=16375, flags=0x00, stream_id=13>
[  0.948] recv WINDOW_UPDATE frame <length=4, flags=0x00, stream_id=13>
          (window_size_increment=16375)
```

**Comment by BH4EHN** [comment]
@yschimke thanks for the real quick response. I do think of this maybe is about the send window, but I have no idea what to do next.

**Comment by yschimke** [comment]
Nginx seems to be really tight with the stream window updates.  Starts with 65KB, doesn't refresh until exhausted and then sends roughly whatever buffer was just sent.  So rather than the stream sitting at 65KB and growing, it gets smaller over time.

Also seems related to what size reads we end up doing from the IO source.  So if you read 8KB chunks we may get stuck on that.

**Comment by BH4EHN** [comment]
@yschimke The service is on ElasticBeanstalk NodeJS environment with Nginx as reverse proxy on every instance, then use ElasticLoadBalancer as load balancer. 
And for the large file upload, I added a file in `.ebextensions` named `instance-proxy.config`
```YAML
files:
  "/etc/nginx/conf.d/proxy.conf" :
    mode: "000755"
    owner: root
    group: root
    content: |
      client_max_body_size 0;
```
And I found if I remove `client_max_body_size 0`, I can get the fast speed (3MB/s). 
May this information helps.
